server/src/test/java/org/opensearch/cluster/routing/MovePrimaryFirstTests.java

```diff
@@ -84,19 +86,24 @@ public void testClusterGreenAfterPartialRelocation() throws InterruptedException
         final ClusterStateListener listener = event -> {
             if (event.routingTableChanged()) {
                 final RoutingNodes routingNodes = event.state().getRoutingNodes();
-                int startedz2n1 = 0;
-                int startedz2n2 = 0;
+                int startedCount = 0;
+                List<ShardRouting> initz2n1 = new ArrayList<>(), initz2n2 = new ArrayList<>();
                 for (Iterator<RoutingNode> it = routingNodes.iterator(); it.hasNext();) {
                     RoutingNode routingNode = it.next();
                     final String nodeName = routingNode.node().getName();
                     if (nodeName.equals(z2n1)) {
-                        startedz2n1 = routingNode.numberOfShardsWithState(ShardRoutingState.STARTED);
+                        startedCount += routingNode.numberOfShardsWithState(ShardRoutingState.STARTED);
+                        initz2n1 = routingNode.shardsWithState(ShardRoutingState.INITIALIZING);
                     } else if (nodeName.equals(z2n2)) {
-                        startedz2n2 = routingNode.numberOfShardsWithState(ShardRoutingState.STARTED);
+                        startedCount += routingNode.numberOfShardsWithState(ShardRoutingState.STARTED);
+                        initz2n2 = routingNode.shardsWithState(ShardRoutingState.INITIALIZING);
                     }
                 }
-                if (startedz2n1 >= primaryShardCount / 2 && startedz2n2 >= primaryShardCount / 2) {
-                    primaryMoveLatch.countDown();
+                if (!Stream.concat(initz2n1.stream(), initz2n2.stream()).anyMatch(s -> s.primary())) {
+                    // All primaries are relocated before 60% of overall shards are started on new nodes
+                    if (primaryShardCount <= startedCount && startedCount <= 6 * primaryShardCount / 5) {
```

Comment:
Maybe it's obvious, but it is not immediately clear to me why the `6 * primaryShardCount / 5` math is correct for calculating that 60% of shards are started on new nodes. Can you explain how this works?

Reply:
Total number of shards are double the primary shard count - `2 * primaryShardCount`. Hence, 60% of total shards is `3 * total number of shards / 5` which is same as `6 * primaryShardCount / 5`

Reply:
Got it. I suggest creating an intermediate variable just to make this more readable, like:

```
final int totalShardCount = primaryShardCount * 2;
if (primaryShardCount <= startedCount && startedCount <= totalShardCount * 3 / 5) {
```

Reply:
Sure, makes sense